Fix teardown for engine's testcases

hubertp opened the pull request:
[ci no changelog needed]

### Pull Request Description

This change intends to cleanup some directories that are being left
behind after running `sbt test`:
- a random `foobar` directory was being created in the `engine` project
directory
- every run of a test suite would add more temporary directories in `/tmp`

The change does not make use of `deleteOnExit` which can be pretty
unreliable. Instead we recursively delete files in directories and
directories to make sure nothing is left behind.
Run scalafmt on changed files to follow the guidelines.

### Important Notes

No architecture changes, just fixing test setup

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

hubertp committed:
Fix teardown for engine's testcases

This changes intends to cleanup some directories that are being left
behind after running `sbt test`:
- a random `foobar` directory was being created in the `engine` project
directory
- every run of a test suite would add more temporary directories in `/tmp`

The change does not make use of `deleteOnExit` which can pretty
unreliable. Instead we recursively delete files in directories and
directories to make sure nothing is left behind.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/enso-org/enso?pullRequest=3303) <br/>All committers have signed the CLA.

wdanilo commented:
Hi @hubertp! Great to see you here! As I do not know this part of the codebase, I believe we have to wait for @4e6 review :)

radeusgd commented on engine/polyglot-api/src/test/scala/org/enso/polyglot/ModuleManagementTest.scala:
```diff
@@ -35,13 +36,27 @@ class ModuleManagementTest extends AnyFlatSpec with Matchers {
 
-  val subject = "Modules Polyglot API"
+    def teardown(): Unit = {
+      Files
+        .walk(tmpDir.toPath())
+        .sorted(Comparator.reverseOrder())
+        .map(_.toFile)
+        .forEach(_.delete())
```
While this looks good, we already have Apache Commons IO in dependencies which provides `FileUtils.deleteDirectory` which serves the same purpose - so maybe it will be better to just use that (especially as some tests are already using it).

```suggestion
      FileUtils.deleteDirectory(tmpDir)
```

(requires also `import org.apache.commons.io.FileUtils`)

hubertp replied:
Cool, I was worried I would be adding a new dependency but that is not the case. Will fix that.

radeusgd reviewed (approved):
Looks good to me, one small suggestion in-line.

We just need to verify that this works robustly on Windows - from the look of it - it should, but sometimes the cleanups were troublesome there (mostly due to Windows holding locks for executables that already terminated for a few miliseconds longer - this issue should not apply in this case, but worth making sure).

For some reason I cannot see the CI statuses - maybe because it's a PR from a fork? @mwu-tow could you have a look to see why the CI doesn't run and what we can do about it?

hubertp committed:
PR review

Use apache's library rather than implementing a custom solution to
delete directories.

mwu-tow commented:
> For some reason I cannot see the CI statuses - maybe because it's a PR from a fork? @mwu-tow could you have a look to see why the CI doesn't run and what we can do about it?

The external forks don't get CI run by design. I've approved it manually.

mwu-tow commented:
@radeusgd 
> We just need to verify that this works robustly on Windows - from the look of it - it should, but sometimes the cleanups were troublesome there (mostly due to Windows holding locks for executables that already terminated for a few miliseconds longer - this issue should not apply in this case, but worth making sure).

Are you sure that this was due to Windows holding lock for terminated executable? I'm not saying that it is false, I just am not aware of such behavior — and this is something that'd be good to know.  Usually lock issues on Windows are due to handles not being closed in an active process — especially because APIs are generally more restrictive (like not being delete running executable or files being opened by default in mode that prevents deleting).

4e6 commented on engine/language-server/src/test/scala/org/enso/languageserver/filemanager/ContentRootManagerSpec.scala:
```diff
@@ -28,12 +29,20 @@ class ContentRootManagerSpec
+
+  var rootManager: ContentRootManagerWrapper = _
+  var rootActor: ActorRef                    = _
+  var rootProjectDirectory: File             = _
```
We have `org.enso.testkit.WithTemporaryDirectory` for these purposes. Please use it instead.

hubertp replied:
done

4e6 commented on engine/polyglot-api/src/test/scala/org/enso/polyglot/ModuleManagementTest.scala:
```diff
@@ -35,13 +36,23 @@ class ModuleManagementTest extends AnyFlatSpec with Matchers {
 
-  val subject = "Modules Polyglot API"
+    def teardown(): Unit = {
+      FileUtils.deleteDirectory(tmpDir)
```
Same. If you're not able to mix the `WithTemporaryDirectory` here, you can extract `robustDeleteDirectory` into an object (make it static) and use it here directly

hubertp replied:
done

radeusgd commented:
> The external forks don't get CI run by design. I've approved it manually.

That's what I thought, thanks for approving it.

> Are you sure that this was due to Windows holding lock for terminated executable? I'm not saying that it is false, I just am not aware of such behavior — and this is something that'd be good to know. Usually lock issues on Windows are due to handles not being closed in an active process — especially because APIs are generally more restrictive (like not being delete running executable or files being opened by default in mode that prevents deleting).

Hmm, the issue I'm mentioning was mostly happening when we were running an executable somewhere (using `java.lang.ProcessBuilder`) and then waiting for it to finish with `waitFor`. IIRC sometimes removing the executable immediately after `waitFor` returned (which should mean that it has already terminated) would fail. But I cannot rule out if some more complicated interactions were going on there.

hubertp committed:
Address PR review

Use `WithTemporaryDirectory` trait that already deals with cleaning up
temporary directories. Only a minor refactoring to the original change
was needed to switch to that logic.

hubertp commented:
I suppose changelog update is not necessary, just not sure where to put `[no-changelog]` so that CI doesn't run the test

radeusgd commented:
> I suppose changelog update is not necessary, just not sure where to put `[no-changelog]` so that CI doesn't run the test

Add `[ci no changelog needed]` at the very top of the PR description. May require pushing some empty commit to re-trigger the check (or maybe just restart of the workflow will do - probably worth trying).

4e6 reviewed (approved):
Awesome!

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/fix-test-project-teardown
